### PR TITLE
Refactor for_display to pass article instead of params

### LINKED ISSUE #19373: Superficial Refactor: Pass an article through to diplay_ads/billboards instead of separate article attributes
## Context

Currently our [`for_display` method in the `DisplayAd` model](https://github.com/forem/forem/blob/8c03359330cf5235ce3e9a5d424c2d30f8af5962/app/models/display_ad.rb#L45-L74) looks as follows: 

```
  def self.for_display(area:, user_signed_in:, organization_id: nil, article_id: nil,
                       article_tags: [], permit_adjacent_sponsors: true)
    ads_for_display = DisplayAds::FilteredAdsQuery.call(
      display_ads: self,
      area: area,
      organization_id: organization_id,
      user_signed_in: user_signed_in,
      article_id: article_id,
      article_tags: article_tags,
      permit_adjacent_sponsors: permit_adjacent_sponsors,
    )

    if rand(8) == 1
      ads_for_display.sample
    else
      ads_for_display.limit(rand(1..15)).sample
    end
  end
```

`article_id`, `article_tags` and `permit_adjacent_sponsors` come from an article resource. From a readability point of view, it may be easier to pass through the whole article instead of these parameters. 

### Requirements

- [ ] Update the `display_for` method to accept an article. Pass it along to the `FilteredAdsQuery`. Extract out the relevant parameters in the `FilteredAdsQuery`. Maybe the extraction should be done in the `display_for` query, open to feedback. 
- [ ] Update all instances where we use `display_for` and `FilteredAdsQuery`
- [ ] Update the relevant tests.

### PR DESCRIPTION
**Pull request by jaw6**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Currently, `DisplayAd.for_display` has parameters which are article attributes (at least: `organization_id`, `article_id`, `tag_list`, and `permit_adjacent_sponsors`). From a readability point of view, it may be easier to pass through the whole article instead of these parameters.

## Related Tickets & Documents

- Closes #19373

## Added/updated tests?

- [x] Yes - refactored to match
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: spec/queries/display_ads/filtered_ads_query_spec.rb [hunk 1]
```diff
@@ -38,13 +39,15 @@ def filter_ads(**options)
     let!(:mismatched) { create_display_ad cached_tag_list: "career" }
 
     it "will show no-tag display ads if the article tags do not contain matching tags" do
-      filtered = filter_ads(article_tags: %w[javascript])
+      allow(article).to receive(:cached_tag_list_array).and_return(%w[javascript])
```
Comment (filleduchaos) [inline spec/queries/display_ads/filtered_ads_query_spec.rb:42]: All the article stubbing feels rather awkward, but I can't really think of a better way to do it

**Review by filleduchaos — APPROVED** [review]
🎉

File: app/queries/display_ads/filtered_ads_query.rb [hunk 2]
```diff
@@ -4,15 +4,18 @@ def self.call(...)
       new(...).call
     end
 
-    def initialize(area:, user_signed_in:, organization_id: nil, article_tags: [],
-                   permit_adjacent_sponsors: true, article_id: nil, display_ads: DisplayAd)
+    # @param area [String] the site area where the ad is visible
+    # @param user_signed_in [Boolean] whether or not the visitor is signed-in
+    # @param article [ArticleDecorator] decorated article is optional
+    # @param display_ads [DisplayAd] can be a filtered scope or Arel relationship
+    def initialize(area:, user_signed_in:, article: nil, display_ads: DisplayAd)
```
Comment (filleduchaos) [inline app/queries/display_ads/filtered_ads_query.rb:11]: Also more YARD docs is always neat! 👍🏾

File: spec/queries/display_ads/filtered_ads_query_spec.rb (hunk 1 above)
Comment (jaw6) [reply spec/queries/display_ads/filtered_ads_query_spec.rb:42]: An option here could be to pass `article` into `DisplayAd.for_display`, but decompose into attributes that are passed into `FilteredAdsQuery` — these tests could then revert back to dealing with an options hash, but we'd still have callers (eg, views/controller) would be simplified. 

Do you think that would make this any better?

Comment (jaw6) [reply spec/queries/display_ads/filtered_ads_query_spec.rb]: @filleduchaos @Ridhwana thoughts on this approach 7c81fbb0 ?

Comment (filleduchaos) [reply spec/queries/display_ads/filtered_ads_query_spec.rb]: Looks good to me! also yay for tiny diffs

**Review by filleduchaos — APPROVED** [review]
🎉

File: spec/queries/display_ads/filtered_ads_query_spec.rb (hunk 1 above)
Comment (Ridhwana) [reply spec/queries/display_ads/filtered_ads_query_spec.rb]: I preferred the previous approach where all the logic and decomposition was encapsulated in the FilteredAdsQuery, but I do see the benefits in simplifying the tests and don't have a super strong opinion on it, so I'm good with this 👍🏽

**Review by Ridhwana — APPROVED** [review]

**Pull request merged**